rpc/lib/client/http_client.go

```diff
@@ -28,67 +28,60 @@ const (
 	protoTCP   = "tcp"
 )
 
-// HTTPClient is a common interface for JSONRPCClient and URIClient.
-type HTTPClient interface {
-	Call(method string, params map[string]interface{}, result interface{}) (interface{}, error)
-	Codec() *amino.Codec
-	SetCodec(*amino.Codec)
+// Parsed URL structure
+type parsedURL struct {
+	url.URL
 }
 
-// protocol - client's protocol (for example, "http", "https", "wss", "ws", "tcp")
-// trimmedS - rest of the address (for example, "192.0.2.1:25", "[2001:db8::1]:80") with "/" replaced with "."
-func toClientAddrAndParse(remoteAddr string) (network string, trimmedS string, err error) {
-	protocol, address, err := parseRemoteAddr(remoteAddr)
+// Parse URL and set defaults (a flexible implementation for 'parseRemoteAddr')
+func newParsedURL(remoteAddr string) (*parsedURL, error) {
+	u, err := url.Parse(remoteAddr)
 	if err != nil {
-		return "", "", err
+		return nil, err
 	}
 
+	// default to tcp if nothing specified
+	if u.Scheme == "" {
+		u.Scheme = protoTCP
+	}
+
+	return &parsedURL{*u}, nil
+}
+
+// Change protocol to HTTP for unknown protocols and TCP protocol (flexible HTTPS downgrade fix) - useful for RPC connections
+func (u *parsedURL) SetDefaultSchemeHTTP() {
```

Comment:
why not do this in `newParsedURL`?

Reply:
Good question, it makes the code even easier. The reason I kept it this way is because I wanted to keep compatibility with the superseded code which originally defaulted to TCP.

I'm not sure if there's any impact if we change this. I propose we change it in the PR for master (I think this will have to be reimplemented there because of other changes). For backward compatibility reasons, I'd leave it like this for this release.

But if you want I'll squash the feature into `newParsedURL`.

Reply:
> I'm not sure if there's any impact if we change this

not sure either lol :) let's not do this here, but do in a separate PR on master. How does that sound?

Reply:
Agreed